esphome/components/sml/sml_parser.cpp

```diff
@@ -0,0 +1,161 @@
+#include "constants.h"
+#include "sml_parser.h"
+
+namespace esphome {
+namespace sml {
+
+SmlBase::SmlBase(const bytes &buffer, unsigned int &pos) : buffer_(buffer), startpos_(pos) {
+  this->type = buffer[pos] >> 4;      // type including overlength info
+  this->length = buffer[pos] & 0x0f;  // length including TL bytes
+  if (this->has_extended_length())    // we have a long list/value (>15 entries)
+    this->length = (this->length << 4) + (buffer[pos + 1] & 0x0f);
+}
+
+bool SmlBase::is_list() { return ((this->type & 0x07) == SML_LIST); }
+
+bool SmlBase::has_extended_length() { return this->type & 0x08; }
+
+SmlNode::SmlNode(const bytes &buffer, unsigned int &pos) : SmlBase(buffer, pos) {
+  uint8_t parse_length = this->length;
+  if (this->has_extended_length()) {
+    pos += 1;
+    parse_length -= 1;
+  }
+
+  if (this->buffer_[pos] == 0x00)  // end of message
+    pos += 1;
+  else if (this->is_list()) {  // list
+    pos += 1;
+    this->nodes.reserve(parse_length);
+    for (unsigned int i = 0; i != parse_length; i++) {
+      this->nodes.emplace_back(SmlNode(this->buffer_, pos));
+    }
+  } else {  // value
+    this->value_bytes = bytes(this->buffer_.begin() + pos + 1, this->buffer_.begin() + pos + parse_length);
+    pos += parse_length;
+  }
+}
+
+SmlFile::SmlFile(bytes buffer) : buffer_(std::move(buffer)) {
+  // extract messages
+  unsigned int pos = 8;
+  while (pos < this->buffer_.size() - 8) {
+    if (this->buffer_[pos] == 0x00)
+      break;  // fill byte detected -> no more messages
+
+    this->messages.emplace_back(SmlNode(this->buffer_, pos));
+  }
+}
+
+std::vector<ObisInfo> SmlFile::get_obis_info() {
+  std::vector<ObisInfo> obis_info;
+  for (auto const &message : messages) {
+    SmlNode message_body = message.nodes[3];
+    uint16_t message_type = bytes_to_uint(message_body.nodes[0].value_bytes);
+    if (message_type != SML_GET_LIST_RES)
+      continue;
+
+    SmlNode get_list_response = message_body.nodes[1];
+    bytes server_id = get_list_response.nodes[1].value_bytes;
+    SmlNode val_list = get_list_response.nodes[4];
+
+    for (auto const &val_list_entry : val_list.nodes) {
+      obis_info.emplace_back(ObisInfo(server_id, val_list_entry));
```

Comment:
```suggestion
      obis_info.emplace_back(server_id, val_list_entry);
```

(`emplace_back` calls the type's constructor, so the `ObisInfo` here is superfluous and would result in an additional move)